### DISCUSSION TITLE
Form Validator validates data twice?

### DISCUSSION THREAD

**Discussion by ghost** [laravel/framework#49911] [Q&A] [answered]
So if Form Validator implements ValidatesWhenResolvedTrait then on resolve from container, it will autovalidate the request.
If request is ok, then if I want to get ->validated() I guess it will revalidate the request a second time. Am I wrong?

Update laravel v 8.

**Comment by rojtjo** [comment]
It reuses the same validator instance, see: https://github.com/laravel/framework/blob/10.x/src/Illuminate/Foundation/Http/FormRequest.php#L234-L244

**Reply by ghost** [reply]
In 8 

    /**
     * Get the validated data from the request.
     *
     * @return array
     */
    public function validated()
    {
        return $this->validator->validated();
    }

which:

    /**
     * Get the attributes and values that were validated.
     *
     * @return array
     *
     * @throws \Illuminate\Validation\ValidationException
     */
    public function validated()


That throw from the dockblock is making me think that it will validate again....

In 10 the same is valid.

**Reply by ghost** [reply]
Just did a test with a unique on db. the query is executed only once so I guess that it is not running the validation twice.

**Reply by rojtjo** [reply, marked as answer]
Yeah I get the confusion here.

When you look at [the implementation](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Validation/Validator.php#L571-L601) you see that it throws an exception when it's invalid:

`throw_if($this->invalid(), $this->exception, $this);`

But when you access the form request from a controller you already know it's valid, so the `invalid` method never returns a thruthy result. When you manually create a validator this method can throw an exception.

I guess that's also the reason why the `validated` method on the form request doesn't have a doc block stating that it throws a `ValidationException`.

**Reply by ghost** [reply]
Thank you

    /**
     * Returns the data which was invalid.
     *
     * @return array
     */
    public function invalid()
    {
        if (! $this->messages) {
            $this->passes();
        }

This if written like this when 

    /**
     * The message bag instance.
     *
     * @var \Illuminate\Support\MessageBag
     */
    protected $messages;

is very confusing. If that property is not set it should throw trying to access property before being initialized. These loose casts to bool are not good practice...

$this->messages->isEmpty();
$this->messages->isNotEmpty();
should have been used if isset ($this->messages)
